README.md
chef-provisioning-opennebula
============================

This is the implementation of the OpenNebula driver for chef-provisioning.  It also comes with additional chef resources to manage OpenNebula:

* Template (one_template)
* Image (one_image)
* VNET (one_vnet)
* Lease (one_vnet_lease)
* OneFlow (one_flow_template & one_flow_service)

Setup
-----
In order to use this driver with chef-provisioning the driver URL needs to be specified via environment variable or ```with_driver``` directive.

Starting in version 0.3.3 the driver URL has the following format:

```ruby
opennebula:<endpoint_url>:<profile>
```

Example:

```ruby
with_driver "opennebula:http://1.2.3.4:443/api:boggi"
```

Where ```boggi``` profile is stored the profile file.  The file will be searched in the following order:
1. ```ENV['ONE_CONFIG']```
2. ```"#{ENV['HOME']}/.one/one_config"```
3. ```/var/lib/one/.one/one_config```

A sample one_config file would look like this:

```json
"boggi": {
  "credentials": "boggi:my_token",
  "options": {
    "timeout": 2,
    "sync": true,
    "http_proxy": "http://my.proxy.com"
  }
},
"testuser": {
  "credentials": "testuser:test_token"
}
```

To use OneFlow, you also need to specify a ```with_flow_url``` directive. This is the API endpoint of your OneFlow.

Example:

```ruby
with_flow_url "http://gary.1.2.3.4:9876"
```

In context of OpenNebula ```machine``` resource will take the following additional options:

```ruby
machine_options {
	:bootstrap_options => {
		:template => Hash defining the VM template,
		:template_file => String location of the VM template file,
		:template_name => String name of the OpenNebula template to use
		:template_id => Integer id of the OpenNebula template to use
		:template_options => Hash values to be merged with VM template
		:is_shutdown => [TrueClass, FalseClass] call vm.shutodwn instead of vm.stop during :stop action
		:shutdown_hard => [TrueClass, FalseClass] flag indicating hard or soft shutdown
    :mode => String octed to set permissions to the machine
	},
	:sudo => true,
	:ssh_username => 'local',
	:ssh_gateway => 'gary@myproxy.net',
	:ssh_options => {
	  Hash containing SSH options as specified by net-ssh gem
	  please see https://github.com/net-ssh/net-ssh for all options
	  Example options:
	  :paranoid => false,
	  :auth_methods => ['publickey'],
	  :keys => [ File.open(File.expand_path('~/.ssh/id_rsa_new')).read ]
	  :keys_only => false,
	  :forward_agent => true,
	  :proxy => 'ssh myproxy.net nc %h %p',
	  :use_agent => true,
	  :user_known_hosts_file => '/dev/null'
	}
  :vm_name => [Symbol, String] Change how the machine shows up in OpenNebula UI and CLI tools.
			       Use :short to rename the machine to the short hostname.
			       Use a string to rename the machine to an arbitrary name.
			       Note this does not change the hostname of the machine, it
			       simply renames the VM in OpenNebula.
  :connection_timeout => [Integer] max wait time to establish connection
}
```

Resources
---------

## one_template

This resource will allow you to create and delete OpenNebula templates.

### Attributes

```ruby
  :template => Hash defining OpenNebula template
  :template_file => String location of the VM template file
  :mode => String octet to set permissions
```

### Actions

```ruby
  actions :create, :delete, :instantiate
  default_action :create
```

### Examples

#### 1. Create OpenNebula template from file

```ruby
one_template "my_one_template" do
    template_file "/opt/one/templates/my_template.tpl"
    action :create
end
```

#### 2. Create OpenNebula template from a template definition

```ruby
one_template "my_one_template" do
  template  ({
    "HTTPBASE" => "http://my.server.com",
    "MEMORY" => "4096",
    "CPU" => "2",
    "VCPU" => "2",
    "OS" => {
        "ARCH" => "x86_64"
    },
    "DISK" => {
        "IMAGE" => "Ubuntu-12.04.5-pre-prod-20141216",
        "IMAGE_UNAME" => "my_user",
        "DRIVER" => "qcow2"
    },
    "NIC" => {
      "NETWORK" => "my_network",
      "NETWORK_UNAME" => "my_network_user"
    },
    "GRAPHICS" => {
          "LISTEN" => "0.0.0.0",
          "TYPE" => "vnc"
    },
    "CONTEXT" => {
          "NETWORK" => "YES",
          "HOSTNAME" => "$NAME",
          "INSTALL_CHEF_CLIENT_COMMAND" => "dpkg -E -i /mnt/chef-client.deb",
          "SSH_USER" => 'local',
          "SSH_PUBLIC_KEY" => "ssh-rsa blahblahblahpublickey root@my_host",
          "FILES" => "$HTTPBASE/01_chef $HTTPBASE/../chef-client.deb"
    }
  })
  action :create
end
```

#### 3. Delete an existing template in OpenNebula

```ruby
one_template "my_one_template" do
    action :delete
end
```

## one_image

This resource will manage images within OpenNebula.  

### Attributes

```ruby
  :name => String name of the image to be created or to be used
  :size => Integer size of the image to allocate in MB
  :datastore_id => Integer id of the datastore in OpenNebula to use 
  :type => ['OS', 'CDROM', 'DATABLOCK', 'KERNEL', 'RAMDISK', 'CONTEXT'] type of the image to create
  :description => String description of the image
  :fs_type => String type of the filesystem to create
  :img_driver => String driver type
  :prefix => String prefix
  :persistent => [ TrueClass, FalseClass] flag indicating if the image should be persistent; default: false
  :public => [ TrueClass, FalseClass ] flag indicating if the image is public
  :disk_type => String Image disk type eq. ext3
  :source => String Image source attribute
  :target => String Image target attribute 
  :image_file => String Local path to the qcow image file. Default: Chef::Config[:file_cache_path]
  :image_id => Integer ID of the image to download
  :download_url => String OpenNebula download URL. Default is ENV['ONE_DOWNLOAD']
  :driver => String Image driver eq. 'qcow2'
  :machine_id => [String, Integer] id of the machine (VM) for disk attach
  :disk_id => [String, Integer] id or name of the disk to attach/snapshot
  :mode => String octet to set permissions
  :http_port => Integer port number to start local HTTP server at, for :image_file uploads. Default: 8066
```

### Actions

```ruby
  actions :allocate, :create, :destroy, :attach, :snapshot, :upload, :download
  default_action :create
```

```ruby
Default attribute values ONLY for image :create and :allocate actions
          :type = 'OS' 
          :fs_type = 'ext2'
          :img_driver = 'qcow2'
          :prefix = 'vd'
          :persistent = false
```

### Examples

#### 1. Create a datablock image with size 10Gb. 

```ruby
one_image "bootstrap-img" do
  type "DATABLOCK"
  size 10240
  datastore_id 103
  action :create
end
```

#### 2. Delete a datablock image with name 'bootstrap-img'

```ruby
one_image "bootstrap-img" do
  action :destroy
end
```

#### 3. Attach a datablock image 'bootstrap-img' to a VM (test-vm) and attach it as 'vde' target.  'target' is optional.

```ruby
one_image "bootstrap-img" do
  machine_id 'test-vm'
  target 'vde'
  action :attach
end
```

#### 4. Take a persistent snapshot of disk 1 on VM (test-vm) and save it as 'snapshot-img'

```ruby
one_image "snapshot-img" do
  machine_id "test-vm"
  disk_id 1
  persistent true
  action :snapshot
end
```

#### 5. Take a persistent snapshot of disk 'bootstrap-img' on VM (test-vm) and save it as 'snapshot-img

```ruby
one_image "snapshot-img" do
  machine_id "test-vm"
  disk_id "bootstrap-img"
  persistent true
  action :snapshot
end
```

#### 6. Upload a local qcow2 image file to OpenNebula, starting HTTP server on port 4567

```ruby
one_image "upload-img" do
  datastore_id 103
  image_file "/local/path/to/qcow/image/file"
  img_driver "qcow2"
  type "OS"
  http_port 4567
  description "This is my cool qcow image"
  action :upload
end
```

#### 7. Upload a qcow2 image file residing on a different host to OpenNebula

```ruby
one_image "upload-img" do
  datastore_id 103
  download_url "http://my.image.host/path/to/qcow/image/file"
  action :upload
end
```

#### 8. Download a 'boggi-test-img' and store it in /home/local/my-image.qcow2.  Download URL is read from ENV[ONE_DOWNLOAD] variable. It will be stored locally in Chef::Config[:file_cache_path]/boggi-test-img.qcow2.

```ruby
one_image "boggi-test-img" do
  action :download
end
```

#### 9. Download image ID 12345 and store it in /tmp/image.qcow2.

```ruby
one_image "boggi-test-img" do
  download_url "http://my.opennebula.download.endpoint/"
  image_id 12345
  image_file "/tmp/image.qcow2"
  action :download
end
```

## one_vnet

This resource will allow to create and delete OpenNebula vnets.

### Attributes

```ruby
  :name => String name of the vnet
  :vnet_id => Integer ID of the vnet that is modified
  :network => Integer ID of the vnet from which a new vnet will be reserved
  :size => Integer size of the new vnet
  :ar_id => Integer address range identifier
  :mac_ip => String ip or mac address
  :template_file => String local file containing the template of a new vnet
  :cluster_id => Integer cluster in which to create a vnet
  :mode => String octet to set permissions
```

### Actions

```ruby
  actions :create, :delete, :reserve
  default_action :reserve
```

### Examples

#### 1. Reserve vnet 'boggi_vnet' from parent vnet 12345 with default size 1

```ruby
one_vnet "boggi_vnet" do
    network 12345
    action :reserve
end
```

#### 2. Reserve vnet 'boggi_vnet' from parent vnet 12345 with size 100

```ruby
one_vnet "boggi_vnet" do
    network 12345
    size 100
    action :reserve
end
```

#### 3. Create a VNET from template file in cluster 12345.  This requires the template file not to have NAME variable populated.  The resource will add NAME attribute to the template with the value being the name of the resource.

```ruby
one_vnet "boggi_vnet" do
    template_file "/tmp/my_vnet.tpl"
    cluster_id 12345
    action :create
end
```

#### 4. Delete a vnet 'boggi_vnet'.  NOTE!!! The vnet cannot have any leases in order to be deleted.

```ruby
one_vnet "boggi_vnet" do
    action :delete
end
```

#### 5. Delete vnet by its ID number (12345).  NOTE!!! The vnet cannot have any leases in order to be deleted.

```ruby
one_vnet "boggi_vnet" do
    vnet_id 12345
    action :delete
end
```

## one_vnet_lease

This resource will allow to hold and release leases within OpenNebula vnets.

### Attributes

```ruby
  :name => String ip or mac address to hold/release
  :vnet_id => Integer ID of the vnet that is modified
  :ar_id => Integer address range identifier, where to allocate the address
  :mac_ip => String ip or mac address to hold/release, same as :name
```

### Actions

```ruby
  actions :hold, :release
  default_action :hold
```

### Examples

#### 1. Hold a lease on a specific IP (1.2.3.4) in 'boggi_vnet' (6789)

```ruby
one_vnet_lease "1.2.3.4" do
    vnet "boggi_vnet"
    action :hold
end
```

#### 2. Hold a lease on a specific MAC (00:00:00:00:01) in a specific address range of 'boggi_vnet' (6789).  Assuming that vnet 6789 has more than one address range.

```ruby
one_vnet_lease "00:00:00:00:01" do
    vnet "boggi_vnet"
    ar_id 1
    action :hold
end
```

#### 3. Release a lease on a specific IP (1.2.3.4) in 'boggi_vnet' (6789). If the IP is already allocated to a VM, that VM must be deleted first, otherwise it will throw an error.

```ruby
one_vnet_lease "1.2.3.4" do
    vnet "boggi_vnet"
    action :release
end
```

## one_user

This resource will allow to create/delete/update OpenNebula users. Right now it does not have the full functionality as `oneuser`, but it can add key=value pairs to an existing user so that sensitive data could be stored there.

### Attributes

```ruby
  :name => String ip or mac address to hold/release
  :password => String password
  :user_id => Integer user id
  :template_file => String local template file defining a user 
  :template => Hash with key/value pairs that will be added to the user
  :append => Boolean append to template or overwrite
```

### Actions

```ruby
  actions :create, :delete, :update
  default_action :update
```

### Examples

#### 1. Create a new user from template file, assuming the current ONE user has the permission to do so.

```ruby
one_user "boggi" do
    template_file "<local_template_file>"
    action :create
end
```

#### 2. Delete 'boggi' user, assuming the current ONE user has the permission to do so.

```ruby
one_user "boggi" do
    action :delete
end
```

#### 3. Add new key/value pairs to user 'boggi', assuming the current ONE user has the permission to do so.

```ruby
one_user "boggi" do
    template ({"BOGGI" => "MAGIC"})
    action :update
end
```

## one_flow_template

This resource will allow you to create, delete, and instantiate OneFlow templates.

### Attributes

```ruby
:name => String name of the OneFlow template you are targeting, default is what you put between 'one_flow_template' and 'do'
:template => Hash defining your OneFlow template, or Fixnum ID of a OneFlow template from ONE, or String file location denoted by file://my_template.json or web location denoted by http(s)://1.2.3.4/my_template.json
:template_options => Hash that override-merges into template
:mode => String octet to set permissions, default is '600'
```

### Actions

```ruby
actions :create, :delete
default_action :create
```

CHANGELOG.md
# Changelog

## 0.4.10 (1/31/2018)
- improve driver checks to allow concurrent running with different versions of OpenNebula

## 0.4.9 (12/05/2017)
- fix rspec tests to work with ONE 4.x and 5.x
- remove dependency upper limit for opennebula gem 

## 0.4.8 (10/24/2016)
- one_image :create action now updates the image if necessary

## 0.4.7 (8/5/2016)
- Added debug info when server_version.rb fails to load 'opennebula' gem, due to bad environment settings
- Added retries to allocate_machine in case ONE has issues and :bootstrap_options are not specified
- Removed default CONTEXT['USER_DATA'] attributes for cloud-init

## 0.4.6 (5/3/2016)
- Added begin/rescue around convergence_strategy deletion in destroy_machine method to avoid failures
  when the client/node objects are not present
- Fixed rubocop errors

## 0.4.5 (4/28/2016)
- Added OneFlow resources (see README): one_flow_template, one_flow_service
- Added RSpec tests for OneFlow
- Fixed RSpec tests for newest chef-client 12.9.38
- Fixed bug regarding destroying machines when machine_spec.reference is nil
- Permitted driver profile switching
- Added backwards compatibility for managing machines provisioned with gem versions prior to v0.3.3
- Added ability to automatically choose the correct opennebula gem version depending on server version

## 0.4.4 (2/29/2016)
- Rewrote RSpec test suite
- Improved chef-run error detection algorithm
- RSpec test results are now organized by datetime and the path of the test recipe
- RSpec now uses `spec/config.rb` rather than environment variables
- Test values, such as disk images, are no longer hardcoded, but are in `config.rb`
- Tests for idempotency are now taken into account
- Expected output for `converge_recipe` can now be specified in regex
- Cleanup recipe has been split into multiple deletion tests
- Cleanup is now run before and after the test suite runs
- Fixed a bug introduced in 0.4.3 which failed to create proper VM template if the primitive values were non-string

## 0.4.3 (2/12/2016)
- Fixed get_pool method to process symbols correctly
- Update driver and resources to use new get_pool method
- Add 4.14 support
- Deprecate :enforce_chef_fqdn
- Added template support for embedded quotation marks
- Added ssh_gateway support
- Add :vm_name option to rename VM in OpenNebula UI

## 0.4.1 (8/12/2015)
- Fixed one_vnet_lease :hold action logic.

## 0.4.0 (7/12/2015)
- Added http_port to one_image resource so that port other than 8066 (default) can be used to upload images.
- Fixed error message in one_image resource
- Removed :credentials and :secret_file support as driver options.  It uses profiles now.
- Fixed ssh timeout.  Now the :timeout will be applied to each ssh command and overall timeout to establish
  a ssh session is 5 min.
- one_image now takes 'download_url' as an optional attribute.  'image_file' still takes precedence over 
  'download_url'.  If 'download_url' is specified then that location is used for image and it will not try 
  to start HTTP server locally.
- Changed default port for HTTP server in one_image to be 8066 instead of previously used port 80.  This is to
  avoid collision with existing web servers running on the host.
- Added :mode attribute to one_image, one_template and one_vnet resources.
- Added bootstrap_options[:mode] so that machines can be created with different permissions

## 0.3.4 (10/29/2015)
- Yanked version 0.3.3 from rubygems due to internal homepage link in gemspec
- Need to release version 0.3.4 to reupload to rubygems.org

## 0.3.3 (10/24/2015)
- Changed driver_url to include profiles.  Without profiles machine_file was unable to recreate the driver
  and subsequently failed. The new driver_url format is:
  opennebula:<endpoint_url>:<profile>
  where profile is stored in ~/.one/one_config or ENV['ONE_CONFIG'] or /var/lib/one/.one/one_config file
- Added check for machine :destroy to verify that the VM is in DONE state.  If not successful after 20 seconds
  it will fail.
- opennebula 4.14 is not backwards compatible so there is a new gem dependency 'opennebula <4.14'.
- TODO: Support for 4.14

## 0.3.2 (10/8/2015)
- Fix rogue property bug

## 0.3.1 (10/5/2015)
- fix bootstrap_options error
- added rubocop test coverage

## 0.3.0 (9/23/2015)
- Added one_user resource
- Added support for machine shutdown.  Before 'machine :stop' would call the stop method on the VM,
  now this behavior can be changed to call vm.shutdown by specifying :bootstrap_options[:is_shutdown] = true
- Added optional bootstrap_option flag [:unique_name] for validation of unique machine names in OpenNebula
- Removed :instantiate from one_template resource, because it is a duplicate or 'machine :template'
- Fixed warnings in providers regarding resource_names
- Removed support for :ssh_execute_options => { :prefix => 'sudo '} in favour of :sudo => true
- Added license headers
- Modified permissions for downloaded qcow images to be 777
- Added error message when :bootstrap_options are not defined
- Added missing :cache attribute to one_image resource


## 0.2.5 (9/10/2015)

- Fixes driver consistency concerns Tyler Ball brought up in [chef-provisioning issue #390](https://github.com/chef/chef-provisioning/issues/390)
- Following machine options are automatically set using these default values unless over written:

```json
:ssh_username => 'local',
    :ssh_options => {
	:keys_only => false,
	:forward_agent => true,
	:use_agent => true,
	:user_known_hosts_file => '/dev/null'
},
:ssh_execute_options => {
:prefix => 'sudo '
}
```

- deprecated machine option :ssh_user, recommend using :ssh_username instead (consistent with aws driver)
- Added ability to set target on :attach action using one_image resource

## 0.2.4 (8/26/2015)

- Added one_vnet and one_vnet_lease resources

## 0.2.2 (7/18/2015)

- Minor fixes

## 0.2.1 (7/11/2015)

- Added upload/download functionality to one_image

## 0.1.0 (4/30/2015)

- Initial submission of chef-provisioning-opennebula gem


lib/chef/provider/one_image.rb
# Copyright 2016, BlackBerry Limited
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

#
# Implementation of Provider class.
#
class Chef
  #
  # Implementation of Provider class.
  #
  class Provider
    #
    # Implementation of Provider class.
    #
    class OneImage < Chef::Provider::LWRPBase
      use_inline_resources

      provides :one_image

      attr_reader :image

      def whyrun_supported?
        true
      end

      def load_current_resource
      end

      def action_handler
        @action_handler ||= Chef::Provisioning::ChefProviderActionHandler.new(self)
      end

      def exists?
        new_driver = driver
        @image = new_driver.one.get_resource(:image, :name => @new_resource.name)
        !@image.nil?
      end

      action :allocate do
        if exists?
          # Image already exists, check whether we need/can update it
          # OpenNebula Image example:
          # ID             : 14202
          # NAME           : ey-test
          # USER           : Mandolin
          # GROUP          : users
          # DATASTORE      : orn-svc01-ds
          # TYPE           : DATABLOCK
          # REGISTER TIME  : 10/19 18:16:02
          # PERSISTENT     : No
          # SOURCE         : /var/lib/one/datastores/103/743dff63f337a0192b13f2963f92d741
          # FSTYPE         : raw
          # SIZE           : 1000M
          # STATE          : rdy
          # RUNNING_VMS    : 0
          #
          # PERMISSIONS
          # OWNER          : um-
          # GROUP          : ---
          # OTHER          : ---
          #
          # IMAGE TEMPLATE
          # DESCRIPTION="what a heck!"
          # DEV_PREFIX="vd"
          # DRIVER="qcow2"
          # FSTYPE="ext4"
          # SIZE="100"
          #
          # We can update many parameters and whatever we update goes into section below 'IMAGE TEMPLATE'.
          # IMPORTANT: if the parameter we are updating exists above 'IMAGE TEMPLATE' then our modification has no effect.
          # In other words a value for a parameter defined above wins over parameter defined below 'IMAGE TEMPLATE'.

          on_image = @image.to_hash['IMAGE']

          # We can update only following attributes. This is a map of one_image attributes to OpenNebula image attributes
          attrs_map = { 'name' => 'NAME', 'size' => 'SIZE', 'datastore_id' => 'DATASTORE_ID', 'type' => 'TYPE', 'description' => 'DESCRIPTION', 'fs_type' => 'FSTYPE',
                        'img_driver' => 'DRIVER', 'prefix' => 'DEV_PREFIX', 'persistent' => 'PERSISTENT', 'mode' => 'PERMISSIONS', 'disk_type' => 'DISK_TYPE' }

          # Find out what attribute needs to be updated
          attrs_to_update = {}
          new_resource_hash = @new_resource.to_hash # hash keys are symbols
          new_resource_hash.each do |k, v|
            next if v.nil? || !attrs_map.key?(k.to_s)
            v = v.to_s # everything is String in what we get in OpenNebula Image info
            on_attr = attrs_map[k.to_s]

            # For some one_image attributes provided in new_resource we need to find respective values in ON Image
            case k
            when :type
              image_types = %w(OS CDROM DATABLOCK KERNEL RAMDISK CONTEXT)
              on_image['TYPE'] = image_types[on_image['TYPE'].to_i] # convert Image Type Id into String
            when :persistent
              on_image['PERSISTENT'] = (on_image['PERSISTENT'] == '1' ? 'true' : 'false')
            when :mode
              perm = on_image['PERMISSIONS']
              perm_octet_u = perm['OWNER_U'].to_i * 4 + perm['OWNER_M'].to_i * 2 + perm['OWNER_A'].to_i
              perm_octet_g = perm['GROUP_U'].to_i * 4 + perm['GROUP_M'].to_i * 2 + perm['GROUP_A'].to_i
              perm_octet_o = perm['OTHER_U'].to_i * 4 + perm['OTHER_M'].to_i * 2 + perm['OTHER_A'].to_i
              on_image['PERMISSIONS'] = "#{perm_octet_u}#{perm_octet_g}#{perm_octet_o}"
            when :disk_type
              disk_types = %w(BLOCK CDROM FILE)
              on_image['DISK_TYPE'] = disk_types[on_image['DISK_TYPE'].to_i] # convert Disk Type into String
            end
            next if on_image.key?(on_attr) && (v == on_image[on_attr])
            next if on_image['TEMPLATE'].key?(on_attr) && (new_resource_hash[k] == on_image['TEMPLATE'][on_attr])
            fail "Cannot update '#{on_attr}' as it is defined above 'IMAGE TEMPLATE' section." if on_image.key?(on_attr) && !on_image[on_attr].empty? && on_attr != 'PERMISSIONS'
            attrs_to_update[on_attr] = v
          end

          unless attrs_to_update.empty?
            # Prepare template to update
            img_template = ''
            attrs_to_update.each do |k, v|
              next if k == 'PERMISSIONS' # needs special treatment
              img_template << case k
                              when 'SIZE', 'PERSISTENT'
                                "#{k} = #{v}\n"
                              when 'TYPE', 'DESCRIPTION', 'FSTYPE', 'DRIVER', 'DEV_PREFIX', 'DISK_TYPE'
                                "#{k} = \"#{v}\"\n"
                              end
            end
            # Perform actual update
            description = "updated image '#{new_resource.name}'\n" + attrs_to_update.to_s
            action_handler.perform_action description do
              unless img_template == '' # can happen when we update only PERMISSIONS
                rc = @image.update(img_template, true)
                fail "failed to update image '#{new_resource.name}': #{rc.message}" if OpenNebula.is_error?(rc)
              end
              if attrs_to_update.key?('PERMISSIONS')
                rc = @image.chmod_octet(attrs_to_update['PERMISSIONS'])
                fail "failed to update image '#{new_resource.name}': #{rc.message}" if OpenNebula.is_error?(rc)
              end
            end
          end
        else
          fail "'size' must be specified" unless new_resource.size
          fail "'datastore_id' must be specified" unless new_resource.datastore_id

          action_handler.perform_action "allocated image '#{new_resource.name}'" do
            @image = new_driver.one.allocate_img(
              :name => new_resource.name,
              :size => new_resource.size,
              :datastore_id => new_resource.datastore_id,
              :type => new_resource.type || 'OS',
              :fs_type => new_resource.fs_type || 'ext2',
              :driver => new_resource.img_driver || 'qcow2',
              :prefix => new_resource.prefix || 'vd',
              :persistent => new_resource.persistent || false)
            new_driver.one.chmod_resource(@image, new_resource.mode)
            Chef::Log.info("Image '#{new_resource.name}' allocate in initial state #{@image.state_str}")
            @new_resource.updated_by_last_action(true)
          end
        end
        @image
      end

      action :create do
        @image = action_allocate
        case @image.state_str
        when 'INIT', 'LOCKED'
          action_handler.perform_action "wait for image '#{new_resource.name}' to be READY" do
            current_driver.one.wait_for_img(new_resource.name, @image.id)
            @new_resource.updated_by_last_action(true)
          end
        when 'READY', 'USED', 'USED_PERS'
        else
          fail "Image #{new_resource.name} is in unexpected state '#{@image.state_str}'"
        end
      end

      action :destroy do
        if exists?
          action_handler.perform_action "deleted image '#{new_resource.name}'" do
            rc = @image.delete
            fail "Failed to delete image '#{new_resource.name}' : #{rc.message}" if OpenNebula.is_error?(rc)
            until new_driver.one.get_resource(:image, :name => new_resource.name).nil?
              Chef::Log.debug("Waiting for delete image to finish...")
              sleep 1
            end
            @new_resource.updated_by_last_action(true)
          end
        else
          action_handler.report_progress "image '#{new_resource.name}' does not exist - (up to date)"
        end
      end

      action :attach do
        fail "Missing attribute 'machine_id'" unless new_resource.machine_id
        fail "Failed to attach disk - image '#{new_resource.name}' does not exist" unless exists?

        vm = new_driver.one.get_resource(:virtualmachine, new_resource.machine_id.is_a?(Integer) ? :id : :name => new_resource.machine_id)
        fail "Failed to attach disk - VM '#{new_resource.machine_id}' does not exist" if vm.nil?
        action_handler.perform_action "attached disk #{new_resource.name} to #{vm.name}" do
          disk_hash = @image.to_hash
          disk_tpl = "DISK = [ "
          disk_tpl << " IMAGE = #{disk_hash['IMAGE']['NAME']}, IMAGE_UNAME = #{disk_hash['IMAGE']['UNAME']}"
          disk_tpl << ", TARGET = #{new_resource.target}" if new_resource.target
          disk_tpl << ", DEV_PREFIX = #{new_resource.prefix}" if new_resource.prefix
          disk_tpl << ", CACHE = #{new_resource.cache}" if new_resource.cache
          disk_tpl << "]"

          disk_id = new_driver.one.get_disk_id(vm, disk_hash['IMAGE']['NAME'])
          if !disk_id.nil?
            action_handler.report_progress "disk is already attached - (up to date)" unless disk_id.nil?
          elsif disk_id.nil?
            action_handler.report_progress "disk not attached, attaching..."
            rc = vm.disk_attach(disk_tpl)
            new_driver.one.wait_for_vm(vm.id)
            fail "Failed to attach disk to VM '#{vm.name}': #{rc.message}" if OpenNebula.is_error?(rc)
            @new_resource.updated_by_last_action(true)
          end
        end
      end

      action :snapshot do
        fail "Missing attribute 'machine_id'" unless new_resource.machine_id
        fail "snapshot '#{new_resource.name}' already exists" if exists?
        vm = new_driver.one.get_resource(:virtualmachine, new_resource.machine_id.is_a?(Integer) ? :id : :name => new_resource.machine_id)
        fail "Failed to create snapshot - VM '#{new_resource.machine_id}' does not exist" if vm.nil?
        action_handler.perform_action "created snapshot from '#{new_resource.machine_id}'" do
          disk_id = new_resource.disk_id.is_a?(Integer) ? new_resource.disk_id : new_driver.one.get_disk_id(vm, new_resource.disk_id)
          fail "No disk '#{new_resource.disk_id}' found on '#{vm.name}'" if disk_id.nil?
          @image = new_driver.one.version_ge_4_14 ? vm.disk_saveas(disk_id, new_resource.name) : vm.disk_snapshot(disk_id, new_resource.name, "", true)
          fail "Failed to create snapshot '#{new_resource.name}': #{@image.message}" if OpenNebula.is_error?(@image)

          @image = new_driver.one.wait_for_img(new_resource.name, @image)
          new_driver.one.chmod_resource(image, new_resource.mode)
          if new_resource.persistent
            action_handler.report_progress "make image '#{new_resource.name}' persistent"
            @image.persistent
          end
          @new_resource.updated_by_last_action(true)
        end
      end

      action :upload do
        fail "'datastore_id' is required" unless new_resource.datastore_id
        fail "'image_file' or 'download_url' attribute is required" unless new_resource.image_file || new_resource.download_url

        file_url = nil
        if new_resource.image_file
          fail "image_file #{new_resource.image_file} does not exist" unless ::File.exist? new_resource.image_file
          file_url = "http://#{node['ipaddress']}:#{@new_resource.http_port}/#{::File.basename(@new_resource.image_file)}"
        else
          file_url = new_resource.download_url
        end
        image_config = {
          :name => @new_resource.name,
          :datastore_id => @new_resource.datastore_id.to_s,
          :path => file_url,
          :driver => @new_resource.img_driver || 'qcow2',
          :description => @new_resource.description || "#{@new_resource.name} image",
          :type => @new_resource.type,
          :mode => @new_resource.mode,
          :prefix => @new_resource.prefix,
          :persistent => @new_resource.persistent,
          :public => @new_resource.public,
          :target => @new_resource.target,
          :disk_type => @new_resource.disk_type,
          :source => @new_resource.source,
          :size => @new_resource.size,
          :fs_type => @new_resource.fs_type
        }

        if exists?
          if @image.name == image_config[:name] &&
             @image['PATH'] == image_config[:path] &&
             @image['TEMPLATE/DRIVER'] == image_config[:driver] &&
             @image['TEMPLATE/DESCRIPTION'] == image_config[:description] &&
             @image['DATASTORE_ID'] == image_config[:datastore_id]
            action_handler.report_progress("image '#{@new_resource.name}' (ID: #{@image.id}) already exists - (up to date)")
          else
            fail "image '#{new_resource.name}' already exists, but it is not the same image"
          end
        else
          action_handler.perform_action "upload image '#{@new_resource.name}'" do
            if @new_resource.image_file
              begin
                success = false
                pid = nil
                trap("CLD") do
                  cpid = Process.wait
                  fail "Could not start HTTP server on port #{@new_resource.http_port}" if cpid == pid && !success
                end
                pid = Process.spawn("python -m SimpleHTTPServer #{@new_resource.http_port}",
                  :chdir => ::File.dirname(@new_resource.image_file),
                  STDOUT => "/dev/null",
                  STDERR => "/dev/null",
                  :pgroup => true)
                new_driver.one.upload_img(image_config)
                success = true
                @new_resource.updated_by_last_action(true)
              ensure
                system("sudo kill -9 -#{pid}")
              end
            else
              new_driver.one.upload_img(image_config)
              @new_resource.updated_by_last_action(true)
            end
          end
        end
      end

      action :download do
        new_driver = driver

        action_handler.perform_action "downloaded image '#{@new_resource.image_file}" do
          download_url = ENV['ONE_DOWNLOAD'] || @new_resource.download_url
          fail %('download_url' is a required attribute.
            You can get the value for 'download_url' by loging into your OpenNebula CLI
            and reading the ONE_DOWNLOAD environment variable) if download_url.nil?
          # You can get the value for 'download_url' by loging into your OpenNebula CLI and reading the ONE_DOWNLOAD environment variable" if download_url.nil?
          image = new_driver.one.get_resource(:image, !@new_resource.image_id.nil? ? { :id => @new_resource.image_id } : { :name => @new_resource.name })
          fail "Image 'NAME: #{@new_resource.name}/ID: #{@new_resource.image_id}' does not exist" if image.nil?
          local_path = @new_resource.image_file || ::File.join(Chef::Config[:file_cache_path], "#{@new_resource.name}.qcow2")
          fail "Will not overwrite an existing file: #{local_path}" if ::File.exist?(local_path)
          command = "curl -o #{local_path} #{download_url}/#{::File.basename(::File.dirname(image['SOURCE']))}/#{::File.basename(image['SOURCE'])}"
          rc = system(command)
          fail rc if rc.nil?
          fail "ERROR: #{rc}" unless rc
          system("chmod 777 #{local_path}")
          Chef::Log.info("Image downloaded from OpenNebula to: #{local_path}")
          @new_resource.updated_by_last_action(true)
        end
      end

      protected

      def driver
        if current_driver && current_driver.driver_url != new_driver.driver_url
          fail "Cannot move '#{machine_spec.name}' from #{current_driver.driver_url} to #{new_driver.driver_url}: machine moving is not supported.  Destroy and recreate."
        end
        fail "Driver not specified for one_image #{new_resource.name}" unless new_driver
        new_driver
      end

      def new_driver
        run_context.chef_provisioning.driver_for(new_resource.driver)
      end

      def current_driver
        run_context.chef_provisioning.driver_for(run_context.chef_provisioning.current_driver) if run_context.chef_provisioning.current_driver
      end
    end
  end
end
